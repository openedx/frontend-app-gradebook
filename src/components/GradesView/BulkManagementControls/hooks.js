import { actions, selectors } from 'data/redux/hooks';

export const useBulkManagementControlsData = () => {
  const gradeExportUrl = selectors.root.useGradeExportUrl();
  const showBulkManagement = selectors.root.useShowBulkManagement();
  const downloadBulkGradesReport = actions.grades.useDownloadBulkGradesReport();

  const handleClickExportGrades = () => {
    downloadBulkGradesReport();
    window.location.assign(gradeExportUrl);
  };

  return {
    show: showBulkManagement,
    handleClickExportGrades,
  };
};
export default useBulkManagementControlsData;
