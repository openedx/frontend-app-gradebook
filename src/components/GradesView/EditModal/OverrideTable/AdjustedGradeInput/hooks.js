import { actions, selectors } from 'data/redux/hooks';
import { getLocalizedSlash } from 'i18n/utils';

const useAdjustedGradeInputData = () => {
  const possibleGrade = selectors.root.useEditModalPossibleGrade();
  const value = selectors.app.useModalData().adjustedGradeValue;
  const setModalState = actions.app.useSetModalState();
  const hintText = possibleGrade && ` ${getLocalizedSlash()} ${possibleGrade}`;

  const onChange = ({ target }) => {
    setModalState({ adjustedGradeValue: target.value });
  };

  return {
    value,
    onChange,
    hintText,
  };
};

export default useAdjustedGradeInputData;
