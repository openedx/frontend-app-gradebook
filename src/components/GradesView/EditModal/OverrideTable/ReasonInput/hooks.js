import React from 'react';

import { actions, selectors } from 'data/redux/hooks';

const useReasonInputData = () => {
  const ref = React.useRef();
  const { reasonForChange } = selectors.app.useModalData();
  const setModalState = actions.app.useSetModalState();

  React.useEffect(() => {
    ref.current.focus();
  }, [ref]);

  const onChange = (event) => {
    setModalState({ reasonForChange: event.target.value });
  };

  return {
    value: reasonForChange,
    onChange,
    ref,
  };
};

export default useReasonInputData;
