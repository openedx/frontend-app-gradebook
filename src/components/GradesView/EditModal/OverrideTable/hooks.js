import { useIntl } from '@edx/frontend-platform/i18n';

import { gradeOverrideHistoryColumns as columns } from 'data/constants/app';
import { selectors } from 'data/redux/hooks';

import messages from './messages';

const useOverrideTableData = () => {
  const { formatMessage } = useIntl();

  const hide = selectors.grades.useHasOverrideErrors();
  const gradeOverrides = selectors.grades.useGradeData().gradeOverrideHistoryResults;
  const tableProps = {};
  if (!hide) {
    tableProps.columns = [
      { Header: formatMessage(messages.dateHeader), accessor: columns.date },
      { Header: formatMessage(messages.graderHeader), accessor: columns.grader },
      { Header: formatMessage(messages.reasonHeader), accessor: columns.reason },
      { Header: formatMessage(messages.adjustedGradeHeader), accessor: columns.adjustedGrade },
    ];
    tableProps.data = gradeOverrides;
  }
  return { hide, ...tableProps };
};

export default useOverrideTableData;
