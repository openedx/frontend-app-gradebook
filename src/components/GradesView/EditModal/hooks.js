import { selectors, actions, thunkActions } from 'data/redux/hooks';

export const useEditModalData = () => {
  const error = selectors.grades.useGradeData().gradeOverrideHistoryError;
  const isOpen = selectors.app.useModalData().open;
  const closeModal = actions.app.useCloseModal();
  const doneViewingAssignment = actions.grades.useDoneViewingAssignment();
  const updateGrades = thunkActions.grades.useUpdateGrades();

  const onClose = () => {
    doneViewingAssignment();
    closeModal();
  };

  const handleAdjustedGradeClick = () => {
    updateGrades();
    doneViewingAssignment();
    closeModal();
  };

  return {
    onClose,
    error,
    handleAdjustedGradeClick,
    isOpen,
  };
};

export default useEditModalData;
