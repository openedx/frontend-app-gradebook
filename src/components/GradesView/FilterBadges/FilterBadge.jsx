import React from 'react';
import PropTypes from 'prop-types';

import { Button } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { selectors } from 'data/redux/hooks';

/**
 * FilterBadge
 * Base filter badge component, that displays a name and a close button.
 * If showValue is true, it will also display the included value.
 * @param {func} handleClose - close/dismiss filter event, taking a list of filternames
 *   to reset when the filter badge closes.
 * @param {string} filterName - api filter name (for redux connector)
 */
export const FilterBadge = ({
  filterName,
  handleClose,
}) => {
  const { formatMessage } = useIntl();
  const {
    displayName,
    isDefault,
    hideValue,
    value,
    connectedFilters,
  } = selectors.root.useFilterBadgeConfig(filterName);
  if (isDefault) {
    return null;
  }
  return (
    <div>
      <span className="badge badge-info">
        <span>{formatMessage(displayName)}</span>
        <span>
          {!hideValue ? `: ${value}` : ''}
        </span>
        <Button
          className="btn-info"
          aria-label="close"
          onClick={handleClose(connectedFilters)}
        >
          <span aria-hidden="true">&times;</span>
        </Button>
      </span>
      <br />
    </div>
  );
};

FilterBadge.propTypes = {
  handleClose: PropTypes.func.isRequired,
  filterName: PropTypes.string.isRequired,
};

export default FilterBadge;
