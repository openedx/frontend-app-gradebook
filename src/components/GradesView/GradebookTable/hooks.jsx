import React from 'react';

import { useIntl } from '@edx/frontend-platform/i18n';

import { selectors } from 'data/redux/hooks';
import transforms from 'data/redux/transforms';
import { Headings } from 'data/constants/grades';
import { getLocalizedPercentSign } from 'i18n/utils';

import messages from './messages';
import Fields from './Fields';
import LabelReplacements from './LabelReplacements';
import GradeButton from './GradeButton';

const { roundGrade } = transforms.grades;

export const useGradebookTableData = () => {
  const { formatMessage } = useIntl();
  const grades = selectors.grades.useAllGrades();
  const headings = selectors.root.useGetHeadings();

  const mapHeaders = (heading) => {
    let label;
    if (heading === Headings.totalGrade) {
      label = <LabelReplacements.TotalGradeLabelReplacement />;
    } else if (heading === Headings.username) {
      label = <LabelReplacements.UsernameLabelReplacement />;
    } else if (heading === Headings.email) {
      label = <LabelReplacements.MastersOnlyLabelReplacement {...messages.emailHeading} />;
    } else if (heading === Headings.fullName) {
      label = <LabelReplacements.MastersOnlyLabelReplacement {...messages.fullNameHeading} />;
    } else {
      label = heading;
    }
    return { Header: label, accessor: heading };
  };

  const mapRows = entry => ({
    [Headings.username]: (
      <Fields.Username username={entry.username} userKey={entry.external_user_key} />
    ),
    [Headings.email]: (<Fields.Text value={entry.email} />),
    [Headings.totalGrade]: `${roundGrade(entry.percent * 100)}${getLocalizedPercentSign()}`,
    ...entry.section_breakdown.reduce((acc, subsection) => ({
      ...acc,
      [subsection.label]: <GradeButton {...{ entry, subsection }} />,
    }), {}),
  });

  const nullMethod = () => null;

  return {
    columns: headings.map(mapHeaders),
    data: grades.map(mapRows),
    grades,
    nullMethod,
    emptyContent: formatMessage(messages.noResultsFound),
  };
};

export default useGradebookTableData;
