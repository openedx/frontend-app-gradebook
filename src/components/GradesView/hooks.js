import { useIntl } from '@edx/frontend-platform/i18n';

import { actions, thunkActions } from 'data/redux/hooks';
import messages from './messages';

export const useGradesViewData = ({ updateQueryParams }) => {
  const { formatMessage } = useIntl();
  const fetchGrades = thunkActions.grades.useFetchGrades();
  const resetFilters = actions.filters.useResetFilters();

  const handleFilterBadgeClose = (filterNames) => () => {
    resetFilters(filterNames);
    updateQueryParams(filterNames.reduce(
      (obj, filterName) => ({ ...obj, [filterName]: false }),
      {},
    ));
    fetchGrades();
  };

  return {
    stepHeadings: {
      filter: formatMessage(messages.filterStepHeading),
      gradebook: formatMessage(messages.gradebookStepHeading),
    },
    handleFilterBadgeClose,
    mastersHint: formatMessage(messages.mastersHint),
  };
};

export default useGradesViewData;
