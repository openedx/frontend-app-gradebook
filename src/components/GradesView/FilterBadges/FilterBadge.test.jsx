import React from 'react';
import { shallow } from 'enzyme';
import { useIntl } from '@edx/frontend-platform/i18n';

import { formatMessage } from 'testUtils';
import { Button } from '@edx/paragon';
import { selectors } from 'data/redux/hooks';
import FilterBadge from './FilterBadge';

jest.mock('@edx/paragon', () => ({
  Button: () => 'Button',
}));
jest.mock('data/redux/hooks', () => ({
  selectors: {
    root: {
      useFilterBadgeConfig: jest.fn(),
    },
  },
}));

const handleClose = jest.fn(filters => ({ handleClose: filters }));
const filterName = 'test-filter-name';

const hookProps = {
  displayName: {
    defaultMessage: 'a common name',
  },
  isDefault: false,
  hideValue: false,
  value: 'a common value',
  connectedFilters: ['some', 'filters'],
};
selectors.root.useFilterBadgeConfig.mockReturnValue(hookProps);

let el;
describe('FilterBadge', () => {
  beforeEach(() => {
    el = shallow(<FilterBadge {...{ handleClose, filterName }} />);
  });
  describe('behavior', () => {
    it('initializes intl hook', () => {
      expect(useIntl).toHaveBeenCalled();
    });
    it('initializes redux hooks', () => {
      expect(selectors.root.useFilterBadgeConfig).toHaveBeenCalledWith(filterName);
    });
  });
  describe('render', () => {
    const testDisplayName = () => {
      test('formatted display name appears on badge', () => {
        expect(el.contains(formatMessage(hookProps.displayName))).toEqual(true);
      });
    };
    const testCloseButton = () => {
      test('close button forwards close method', () => {
        expect(el.find(Button).props().onClick).toEqual(handleClose(hookProps.connectedFilters));
      });
    };
    test('empty render if isDefault', () => {
      selectors.root.useFilterBadgeConfig.mockReturnValueOnce({
        ...hookProps,
        isDefault: true,
      });
      el = shallow(<FilterBadge {...{ handleClose, filterName }} />);
      expect(el.isEmptyRender()).toEqual(true);
    });
    describe('hide Value', () => {
      beforeEach(() => {
        selectors.root.useFilterBadgeConfig.mockReturnValueOnce({
          ...hookProps,
          hideValue: true,
        });
        el = shallow(<FilterBadge {...{ handleClose, filterName }} />);
      });
      testDisplayName();
      testCloseButton();
      test('snapshot', () => {
        expect(el).toMatchSnapshot();
      });
      test('value is note present in the badge', () => {
        expect(el.contains(hookProps.value)).toEqual(false);
      });
    });
    describe('do not hide value', () => {
      testDisplayName();
      testCloseButton();
      test('snapshot', () => {
        expect(el).toMatchSnapshot();
      });
      test('value is note present in the badge', () => {
        expect(el.text().includes(hookProps.value)).toEqual(true);
      });
    });
  });
});
