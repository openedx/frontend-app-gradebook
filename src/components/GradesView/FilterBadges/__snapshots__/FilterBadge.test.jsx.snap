// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBadge render do not hide value snapshot 1`] = `
<div>
  <span
    className="badge badge-info"
  >
    <span>
      a common name
    </span>
    <span>
      : a common value
    </span>
    <Button
      aria-label="close"
      className="btn-info"
      onClick={
        Object {
          "handleClose": Array [
            "some",
            "filters",
          ],
        }
      }
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </Button>
  </span>
  <br />
</div>
`;

exports[`FilterBadge render hide Value snapshot 1`] = `
<div>
  <span
    className="badge badge-info"
  >
    <span>
      a common name
    </span>
    <span />
    <Button
      aria-label="close"
      className="btn-info"
      onClick={
        Object {
          "handleClose": Array [
            "some",
            "filters",
          ],
        }
      }
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </Button>
  </span>
  <br />
</div>
`;
